javascript/react-js/state-and-props.md

```diff
@@ -0,0 +1,240 @@
+
+Another very important concept in React is state. In short, all values that can change over time are handled through state.
+
+And that's the way to define `state` in React:
+
+```javascript
+import React, { Components } from "react";
+
+class App extends Component {
+  constructor() {
+    super();
+
+    this.state = {
+      count: 0,
+    };
+  }
+
+  countUp() {
+    this.setState({
+      count: this.state.count + 1,
+    });
+  }
+
+  render() {
+    return (
+      <div>
+        <button onClick={this.countUp}>Click Me!</button>
+        <p>{this.state.count}</p>
+      </div>
+    );
+  }
+}
+```
+
+In the above component we declared a state. You **always** declare state in the constructor in a class component. Remember that for functional components it works a little bit different. The `setState` method we call sets the state to a new value. In React you should **never** change state directly. What does that mean? Never do something like this: `this.state.count = 3`, instead always use the `setState` method React provides you to modify the state. Really keep that in mind, it can save you a lot of debugging when you are getting started with React.
+
+As you have also seen in the `render` method, we access the current state through `this.state.count`. This syntax should look familiar to you by now, because it is the same way we accessed props. And yes, you can also destructure state.
+
+### State and props in functional components
+
+As you have learned already, React also provides the ability to create functional components. And well, they work a little bit different than the class components we discussed most of the time. The concept of props basically works the same, with the only difference, you don't pass `props` as an argument to the constructor, but instead you pass it as an argument to the component itself. And another major difference between functional and class components concerning props is the way you call the props. You learned that in a class component the props that have been passed down from the parent component will be called like so: `this.props.someFunction`, however in functional components, we don't need to call `this`, so we access `props` like so: `props.someFunction`. And that's already the whole difference about `props` between class and functional components. Let me give you a quick example to emphasise this.
+
+```javascript
+// MyComponent.js
+
+import React from "react";
+
+const MyComponent = (props) => {
+  return <div>{props.title}</div>;
+};
+
+export default MyComponent;
+```
+
+```javascript
+// App.js
+
+import React from "react";
+import MyComponent from "./MyComponent";
+
+const App = () => {
+  return (
+    <div>
+      <MyComponent title="Hello World" />
+    </div>
+  );
+};
+
+export default App;
+```
+
+Of course we can also use destructuring here, by adding this line `const {title} = props` above the return statement in `MyComponent.js` we can just refere to the title using `{title}`
```

Comment:
We should explicitly add destructuring props when there are several present.

Reply:
Unfortunately, I am not sure what you mean by that.  I changed the sentence to: "Of course we can also destructure from props here. By adding ..." I hope that's what you meant.

Reply:
That's perfectly fine.